In [5]:
'''Q1 Dataset Exploration

Perform the following exploratory steps:

Display the first 10 rows

Display the dataset structure using info()

Find the number of rows and columns

Identify the data types of each column
'''
import pandas as pd
df =pd.read_csv("Advanced_pandas_sales_dataset.csv")
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [12]:
df.dtypes

,0
Order_ID,int64
Order_Date,object
Customer,object
City,object
Region,object
Product,object
Category,object
Sales,int64
Quantity,int64
Discount,float64


In [7]:
df.shape

(230, 11)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


In [13]:
'''Q2 Sales Filtering Analysis

Filter the dataset to display:

Orders where Sales > 3500

Orders where Region = "East" and Sales > 2500

Explain what insights you observe.

'''
df[df["Sales"]>3500]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
10,1011,2024-06-30,Neha,Delhi,South,Printer,Office,5977,6,0.00,1578.32
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


In [20]:
df[(df["Region"]=="East") &( df["Sales"]>2500)]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
34,1035,2024-02-20,Nitin,Ahmedabad,East,Laptop,Electronics,2535,4,0.10,188.15
41,1042,2024-07-15,Rohan,Chennai,East,Laptop,Electronics,3268,8,0.15,295.15
43,1044,2024-09-20,Kavya,Hyderabad,East,Printer,Office,2618,7,0.05,454.64


Q2 **insight**

**Insight: The filtered data shows that the East region has many high-value sales orders, with sales above 2500. Several orders have sales above 3500, and Electronics products appear frequently in the filtered results. Most of these orders have positive profit, although a few high-sales orders have negative profit.**

In [27]:
'''Q3 Customer Purchase Analysis

Identify the top 5 customers based on total sales.

Hint: Use GroupBy with aggregation.
'''
df.groupby("Customer")["Sales"].sum().sort_values(ascending=False).head()

,Sales
Customer,
Nitin,59820
Raj,58416
Rahul,52291
Dev,48500
Isha,46818


In [29]:
df.groupby("Region")["Sales"].sum().sort_values(ascending=False)

,Sales
Region,
East,232675
South,183108
West,170227
North,158210


In [30]:
df.groupby("Region")["Sales"].sum().idxmax()

'East'

In [31]:
'''Q5 Product Performance Analysis

Calculate the total quantity sold for each product.

Which product appears to be the most popular?
'''
df.groupby("Product")["Quantity"].sum()

,Quantity
Product,
Headphones,115
Keyboard,110
Laptop,140
Monitor,156
Mouse,122
Phone,151
Printer,108
Tablet,150


In [32]:
df.groupby("Product")["Quantity"].sum().sort_values(ascending=False)

,Quantity
Product,
Monitor,156
Phone,151
Tablet,150
Laptop,140
Mouse,122
Headphones,115
Keyboard,110
Printer,108


In [33]:
df.groupby("Product")["Quantity"].sum().idxmax()

'Monitor'

In [35]:
'''Q6 Multi-Level Grouping

Group the dataset by: Region and Category
Calculate the total sales for each combination.

What insights can you derive from this analysis?

'''
df.groupby(["Region","Category"])["Sales"].sum().sort_values(ascending=False)

,,Sales
Region,Category,
East,Electronics,160907
South,Electronics,86690
West,Electronics,84128
South,Accessories,80914
North,Electronics,70023
West,Accessories,64811
North,Accessories,62052
East,Accessories,57646
North,Office,26135


Q6 (2)**insigh**t: The East region has the highest sales for the Electronics category, with total sales of 160,907. The Office category has comparatively lower sales across all regions, with East having the lowest at 14,122.


In [37]:
'''Q7 Pivot Table Creation

Create a pivot table showing:

Rows → Region

Columns → Product

Values → Total Sales

Which region sells the most laptops?
'''
pd.pivot_table(df,index="Region",columns="Product",values="Sales",aggfunc="sum")

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


**Q7(2)-East region sells the most laptops, with total sales of 55,568.**

In [40]:
'''Q8 Profitability Analysis

Calculate the average profit for each product category.

Which category is most profitable?
'''
df.groupby("Category")["Profit"].mean()

,Profit
Category,
Accessories,410.603086
Electronics,396.462114
Office,471.938462


In [41]:
df.groupby("Category")["Profit"].mean().sort_values(ascending=False)

,Profit
Category,
Office,471.938462
Accessories,410.603086
Electronics,396.462114


In [42]:
'''Q9 Discount Impact Study

Analyze the relationship between discount and profit.

Tasks:

Calculate the average profit for each discount level

Identify whether higher discounts lead to lower profit
'''
df[["Discount","Profit"]]

,Discount,Profit
0,0.00,307.52
1,0.15,49.89
2,0.20,66.47
3,0.20,-71.66
4,0.20,146.30
...,...,...
225,0.15,662.00
226,0.05,323.13
227,0.15,119.82
228,0.00,531.76


In [43]:
df.groupby("Discount")["Profit"].mean()

,Profit
Discount,
0.00,751.210588
0.05,531.368039
0.10,346.460513
0.15,237.191400
0.20,90.026923


**Q9(2)-The data shows that higher discounts are associated with lower average profit. As the discount increases from 0% to 20%, the average profit decreases from approximately 751.21 to 90.03**

In [51]:
'''Q10 End-to-End Data Wrangling Task

Perform the following workflow:

Filter the dataset where Sales > 3000

Group data by Region and Product

Calculate total sales

Reshape the result into a pivot table

Explain the business insights obtained from this analysis.
'''
filtered_df = df[df["Sales"] > 3000]
grouped_sales = filtered_df.groupby(["Region", "Product"])["Sales"].sum()

In [49]:
df.groupby(["Region","Product"])["Sales"].sum()

Region  Product   
East    Headphones    23899
        Keyboard      10845
        Laptop        55568
        Monitor       31299
        Mouse         22902
        Phone         39667
        Printer       14122
        Tablet        34373
North   Headphones    15450
        Keyboard      20538
        Laptop         4540
        Monitor       14922
        Mouse         26064
        Phone         30535
        Printer       26135
        Tablet        20026
South   Headphones    24111
        Keyboard      23480
        Laptop        23029
        Monitor       19167
        Mouse         33323
        Phone         18218
        Printer       15504
        Tablet        26276
West    Headphones    17159
        Keyboard      38499
        Laptop        10826
        Monitor       18178
        Mouse          9153
        Phone         27850
        Printer       21288
        Tablet        27274
Name: Sales, dtype: int64

In [52]:
pd.pivot_table(
    filtered_df,
    index="Region",
    columns="Product",
    values="Sales",
    aggfunc="sum"
)

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,14469,9642,44096,16420,19000,29500,3696,29229
North,5326,15653,4540,5087,22209,24670,17185,3392
South,10026,19249,17950,14921,26928,18218,14143,19613
West,13898,33380,5922,14354,3511,15653,12648,17185


In [54]:
pivot_table = grouped_sales.unstack()
pivot_table

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,14469,9642,44096,16420,19000,29500,3696,29229
North,5326,15653,4540,5087,22209,24670,17185,3392
South,10026,19249,17950,14921,26928,18218,14143,19613
West,13898,33380,5922,14354,3511,15653,12648,17185


**Q10(5)Business Insight: The filtered data shows different sales patterns across regions and products. East has high sales for Laptop and Phone, while West has particularly high sales for Keyboard. North has the highest sales for Printer within that region. These differences can help businesses understand which products perform strongly in different regions **bold text**